zone/doors.cpp

```diff
@@ -190,6 +190,204 @@ bool Doors::Process()
+					instanceLockout.character_id = sender->CharacterID();
+					instanceLockout.expirydate = end_time;
+					instanceLockout.zone_id = zoneid;
+					instanceLockout.zone_instance_id = zoneguildid;
+					sender->character_instance_lockouts[zoneid] = instanceLockout;
+				}
+			}
+			else
+			{
+				sender->Message(Chat::Red, "You are unable to enter an instance because your raid leader hasn't entered first.");
+				return false;
+			}
+		}
+		else
+		{
+			uint32 leader_instanceid = zone->GetZoneInstanceIDByCharacterAndZone(player_raid->GetRaidLeaderCharacterID(), zoneid);
+			if (ourZoneInstanceID == GUILD_NONE)
+			{
+
+				zoneguildid = database.GetHighestZoneInstanceID(zoneid) + 1;
+				//TODO: hardcode, retrieve from db later
+				int64 zonelockout = RuleI(Quarm, InstanceMinimumLockoutTime);
+				auto cur_time = time(nullptr);
+				int64 end_time = cur_time + RuleI(Quarm, InstanceMinimumLockoutTime);
+				int64 curzoneLockout = zone->GetZoneInstanceLockoutByCharacterAndZone(sender->CharacterID(), zoneid);
+				if (curzoneLockout <= end_time)
+				{
+					zone->ReplaceZoneInstanceIDCache(sender->CharacterID(), zoneid, zoneguildid, end_time);
+					database.SaveCharacterInstanceLockout(sender->CharacterID(), end_time, zoneid, zoneguildid);
+					CharacterInstanceLockout instanceLockout;
+					memset(&instanceLockout, 0, sizeof(CharacterInstanceLockout));
+					instanceLockout.character_id = sender->CharacterID();
+					instanceLockout.expirydate = end_time;
+					instanceLockout.zone_id = zoneid;
+					instanceLockout.zone_instance_id = zoneguildid;
+					sender->character_instance_lockouts[zoneid] = instanceLockout;
+				}
+			}
+		}
+	}
+
+	if (zoneguildid == GUILD_NONE)
+	{
+		sender->Message(Chat::Red, "You are unable to enter this instance.");
+		return false;
+	}
+
+	if (strncmp(destination_zone_name, zone_name, strlen(zone_name)) == 0 && !keyneeded)
+	{
+		if (!keepoffkeyring)
+		{
+			sender->KeyRingAdd(key);
+		}
+		sender->MovePCGuildID(zone->GetZoneID(), zoneguildid, m_destination.x, m_destination.y, m_destination.z, m_destination.w);
+	}
+	else if ((!IsDoorOpen() || open_type == 58) && (keyneeded && ((keyneeded == key) || sender->GetGM())))
+	{
+		if (!keepoffkeyring)
+		{
+			sender->KeyRingAdd(key);
+		}
+		if (zoneid == zone->GetZoneID())
+		{
+			sender->MovePCGuildID(zone->GetZoneID(), zoneguildid, m_destination.x, m_destination.y, m_destination.z, m_destination.w);
+		}
+		else
+		{
+			zone->ApplyRandomLoc(zoneid, temp_x, temp_y);
+			sender->MovePCGuildID(zoneid, zoneguildid, temp_x, temp_y, m_destination.z, m_destination.w);
+		}
+	}
+
+	if ((!IsDoorOpen() || open_type == 58) && !keyneeded)
+	{
+		if (zoneid == zone->GetZoneID())
+		{
+			sender->MovePCGuildID(zone->GetZoneID(), zoneguildid, m_destination.x, m_destination.y, m_destination.z, m_destination.w);
+		}
+		else
+		{
+			zone->ApplyRandomLoc(zoneid, temp_x, temp_y);
+			sender->MovePCGuildID(zoneid, zoneguildid, temp_x, temp_y, m_destination.z, m_destination.w);
```

Comment:
Since the MovePCGuildID logic is repeated several times, I wonder if a better pattern for future maintenance might be to set up a couple bools in the above if/elses for if the MovePCGuildID should be called & if a random loc should be used